### PR TITLE
Add discovery request table entry on discovery request. Remove discovery request on successful link.

### PR DESCRIPTION
**Pull request by chethandeshpande**
Discovery and link to have separate database context files. 
Moved migrations into respective folders.
Database migrations to run on Startup.
Add discovery request entry on discovery request
Remove discovery request on successful link.

### CODE REVIEW

File: hip-service/Discovery/Patient/Database/Migrations/20200116095059_AddDiscoveryRequestTable.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
+﻿using System;
+using Microsoft.EntityFrameworkCore.Migrations;
+using Npgsql.EntityFrameworkCore.PostgreSQL.Metadata;
+
+namespace hip_service.Discovery.Patient.Database.Migrations
+{
+    public partial class AddDiscoveryRequestTable : Migration
```
Comment (ganesanarun) [inline hip-service/Discovery/Patient/Database/Migrations/20200116095059_AddDiscoveryRequestTable.cs]: Codacy Issue found: ['partial' is gratuitous in this context.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4833676)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- hip-service-test/Discovery/Patient/PatientDiscoveryTest.cs  1
- hip-service/Link/Patient/Database/Migrations/20200110085309_intial.cs  1
- hip-service-test/Discovery/Patient/DiscoveryRequestRepositoryTest.cs  1
- hip-service/Discovery/Patient/Database/DiscoveryContext.cs  1
- hip-service/Link/Patient/Database/Migrations/LinkPatientContextModelSnapshot.cs  1
- hip-service/Discovery/Patient/Database/Migrations/DiscoveryContextModelSnapshot.cs  1
- hip-service/Discovery/Patient/Database/Migrations/20200116095059_AddDiscoveryRequestTable.cs  1
- hip-service/Discovery/Patient/Model/DiscoveryRequest.cs  1
- hip-service/Link/Patient/Database/LinkPatientContext.cs  1
- hip-service/Link/Patient/Database/Migrations/20200110085309_intial.Designer.cs  1
- hip-service/Discovery/Patient/DiscoveryRequestRepository.cs  2
- hip-service/Discovery/Patient/Database/Migrations/20200116095059_AddDiscoveryRequestTable.Designer.cs  1
- hip-service/Discovery/Patient/PatientDiscovery.cs  1
         

Clones added
============
- hip-service/Link/Patient/Database/Migrations/LinkPatientContextModelSnapshot.cs  1
- hip-service/Discovery/Patient/Database/Migrations/DiscoveryContextModelSnapshot.cs  1
- hip-service/Link/Patient/Database/Migrations/20200110085309_intial.Designer.cs  1
- hip-service/Discovery/Patient/Database/Migrations/20200116095059_AddDiscoveryRequestTable.Designer.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4833676&bid=16055942)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/In.ProjectEKA.HipService/Discovery/Patient/PatientDiscovery.cs  1
- test/In.ProjectEKA.HipServiceTest/Discovery/Patient/DiscoveryRequestRepositoryTest.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/DiscoveryRequestRepository.cs  2
- test/In.ProjectEKA.HipServiceTest/Discovery/Patient/PatientDiscoveryTest.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/Database/DiscoveryContext.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/Database/Migrations/20200116095059_AddDiscoveryRequestTable.Designer.cs  1
- src/In.ProjectEKA.HipService/Link/Patient/Database/Migrations/20200110085309_intial.Designer.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/Model/DiscoveryRequest.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/Database/Migrations/20200116095059_AddDiscoveryRequestTable.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/Database/Migrations/DiscoveryContextModelSnapshot.cs  1
- src/In.ProjectEKA.HipService/Link/Patient/Database/LinkPatientContext.cs  1
- src/In.ProjectEKA.HipService/Link/Patient/Database/Migrations/20200110085309_intial.cs  1
- src/In.ProjectEKA.HipService/Link/Patient/Database/Migrations/LinkPatientContextModelSnapshot.cs  1
         

Clones added
============
- src/In.ProjectEKA.HipService/Discovery/Patient/Database/Migrations/20200116095059_AddDiscoveryRequestTable.Designer.cs  1
- src/In.ProjectEKA.HipService/Link/Patient/Database/Migrations/20200110085309_intial.Designer.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/Database/Migrations/DiscoveryContextModelSnapshot.cs  1
- src/In.ProjectEKA.HipService/Link/Patient/Database/Migrations/LinkPatientContextModelSnapshot.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4833676&bid=16055942)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/In.ProjectEKA.HipService/Discovery/Patient/Database/Migrations/20200116095059_AddDiscoveryRequestTable.Designer.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/Database/Migrations/DiscoveryContextModelSnapshot.cs  1
- src/In.ProjectEKA.HipService/Link/Patient/Database/Migrations/20200110085309_intial.cs  1
- test/In.ProjectEKA.HipServiceTest/Discovery/Patient/DiscoveryRequestRepositoryTest.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/Database/DiscoveryContext.cs  1
- src/In.ProjectEKA.HipService/Link/Patient/Database/LinkPatientContext.cs  1
- src/In.ProjectEKA.HipService/Link/Patient/Database/Migrations/20200110085309_intial.Designer.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/Model/DiscoveryRequest.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/DiscoveryRequestRepository.cs  2
- src/In.ProjectEKA.HipService/Discovery/Patient/PatientDiscovery.cs  1
- test/In.ProjectEKA.HipServiceTest/Discovery/Patient/PatientDiscoveryTest.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/Database/Migrations/20200116095059_AddDiscoveryRequestTable.cs  1
- src/In.ProjectEKA.HipService/Link/Patient/Database/Migrations/LinkPatientContextModelSnapshot.cs  1
         

Clones added
============
- src/In.ProjectEKA.HipService/Discovery/Patient/Database/Migrations/20200116095059_AddDiscoveryRequestTable.Designer.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/Database/Migrations/DiscoveryContextModelSnapshot.cs  1
- src/In.ProjectEKA.HipService/Link/Patient/Database/Migrations/20200110085309_intial.Designer.cs  1
- src/In.ProjectEKA.HipService/Link/Patient/Database/Migrations/LinkPatientContextModelSnapshot.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4833676&bid=16055942)

**Comment by ganesanarun** [comment]
Make sure migration works on empty database before commit the change.

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/In.ProjectEKA.HipService/Link/Patient/Database/Migrations/LinkPatientContextModelSnapshot.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/Model/DiscoveryRequest.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/Database/Migrations/20200116095059_AddDiscoveryRequestTable.Designer.cs  1
- src/In.ProjectEKA.HipService/Link/Patient/Database/LinkPatientContext.cs  1
- src/In.ProjectEKA.HipService/Link/Patient/Database/Migrations/20200110085309_intial.Designer.cs  1
- test/In.ProjectEKA.HipServiceTest/Discovery/Patient/DiscoveryRequestRepositoryTest.cs  1
- test/In.ProjectEKA.HipServiceTest/Discovery/Patient/PatientDiscoveryTest.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/DiscoveryRequestRepository.cs  2
- src/In.ProjectEKA.HipService/Discovery/Patient/Database/DiscoveryContext.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/PatientDiscovery.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/Database/Migrations/20200116095059_AddDiscoveryRequestTable.cs  1
- src/In.ProjectEKA.HipService/Link/Patient/Database/Migrations/20200110085309_intial.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/Database/Migrations/DiscoveryContextModelSnapshot.cs  1
         

Clones added
============
- src/In.ProjectEKA.HipService/Link/Patient/Database/Migrations/LinkPatientContextModelSnapshot.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/Database/Migrations/20200116095059_AddDiscoveryRequestTable.Designer.cs  1
- src/In.ProjectEKA.HipService/Link/Patient/Database/Migrations/20200110085309_intial.Designer.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/Database/Migrations/DiscoveryContextModelSnapshot.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4833676&bid=16055942)

File: test/In.ProjectEKA.HipServiceTest/Discovery/Patient/DiscoveryRequestRepositoryTest.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+            await discoveryRequestRepository.Add(discoveryRequest);
+
+            count = await dbContext.DiscoveryRequest.CountAsync();
+            count.Should().Be(1);
+
+            await discoveryRequestRepository.Delete(transactionId, patientId);
+
+            count = await dbContext.DiscoveryRequest.CountAsync();
+            count.Should().Be(0);
```
[the remark this answers is not in the document]
Comment (chethandeshpande) [reply test/In.ProjectEKA.HipServiceTest/Discovery/Patient/DiscoveryRequestRepositoryTest.cs]: I had two tests initially. But then to test delete, it needs to be added. Thats why I chose to keep one test.

File: src/In.ProjectEKA.HipService/Startup.cs [hunk 3]
```diff
@@ -54,6 +57,12 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
                 .UseCustomOpenAPI()
                 .UseConsentManagerIdentifierMiddleware()
                 .UseEndpoints(endpoints => { endpoints.MapControllers(); });
+
+            using var serviceScope = app.ApplicationServices.GetRequiredService<IServiceScopeFactory>().CreateScope();
```
[the remark this answers is not in the document]
Comment (chethandeshpande) [reply src/In.ProjectEKA.HipService/Startup.cs]: It is Disposable object

File: src/In.ProjectEKA.HipService/Discovery/Patient/PatientDiscovery.cs [hunk 4]
```diff
@@ -11,23 +11,28 @@ namespace In.ProjectEKA.HipService.Discovery.Patient
...
         {
             var expression = GetExpression(request.Patient.VerifiedIdentifiers);
-            var patientInfos = await repo.Where(expression);
+            var patientInfos = await matchingRepository.Where(expression);
             var (patient, error) = DiscoveryUseCase.DiscoverPatient(filter.Do(patientInfos, request).AsQueryable());
 
-            return new Tuple<DiscoveryResponse, ErrorResponse>(
-                new DiscoveryResponse(patient),
-                error);
+            if (patient == null) return new Tuple<DiscoveryResponse, ErrorResponse>(null, error);
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/Discovery/Patient/PatientDiscovery.cs]: Codacy Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4833676)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- src/In.ProjectEKA.HipService/Discovery/Patient/Database/DiscoveryContext.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/Database/Migrations/20200116095059_AddDiscoveryRequestTable.Designer.cs  1
- src/In.ProjectEKA.HipService/Link/Patient/Database/Migrations/20200110085309_intial.cs  1
- test/In.ProjectEKA.HipServiceTest/Discovery/Patient/PatientDiscoveryTest.cs  1
- src/In.ProjectEKA.HipService/Link/Patient/Database/Migrations/LinkPatientContextModelSnapshot.cs  1
- src/In.ProjectEKA.HipService/Link/Patient/Database/LinkPatientContext.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/PatientDiscovery.cs  1
- test/In.ProjectEKA.HipServiceTest/Discovery/Patient/DiscoveryRequestRepositoryTest.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/DiscoveryRequestRepository.cs  2
- src/In.ProjectEKA.HipService/Link/Patient/Database/Migrations/20200110085309_intial.Designer.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/Database/Migrations/20200116095059_AddDiscoveryRequestTable.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/Database/Migrations/DiscoveryContextModelSnapshot.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/Model/DiscoveryRequest.cs  1
         

Clones added
============
- src/In.ProjectEKA.HipService/Discovery/Patient/Database/Migrations/20200116095059_AddDiscoveryRequestTable.Designer.cs  1
- src/In.ProjectEKA.HipService/Link/Patient/Database/Migrations/LinkPatientContextModelSnapshot.cs  1
- src/In.ProjectEKA.HipService/Link/Patient/Database/Migrations/20200110085309_intial.Designer.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/Database/Migrations/DiscoveryContextModelSnapshot.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4833676&bid=16055942)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/In.ProjectEKA.HipService/Discovery/Patient/Database/DiscoveryContext.cs  1
- src/In.ProjectEKA.HipService/Link/Patient/Database/Migrations/LinkPatientContextModelSnapshot.cs  1
- src/In.ProjectEKA.HipService/Link/Patient/Database/Migrations/20200110085309_intial.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/DiscoveryRequestRepository.cs  2
- test/In.ProjectEKA.HipServiceTest/Discovery/Patient/PatientDiscoveryTest.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/PatientDiscovery.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/Database/Migrations/20200116095059_AddDiscoveryRequestTable.Designer.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/Model/DiscoveryRequest.cs  1
- test/In.ProjectEKA.HipServiceTest/Discovery/Patient/DiscoveryRequestRepositoryTest.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/Database/Migrations/DiscoveryContextModelSnapshot.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/Database/Migrations/20200116095059_AddDiscoveryRequestTable.cs  1
- src/In.ProjectEKA.HipService/Link/Patient/Database/LinkPatientContext.cs  1
- src/In.ProjectEKA.HipService/Link/Patient/Database/Migrations/20200110085309_intial.Designer.cs  1
         

Clones added
============
- src/In.ProjectEKA.HipService/Link/Patient/Database/Migrations/LinkPatientContextModelSnapshot.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/Database/Migrations/20200116095059_AddDiscoveryRequestTable.Designer.cs  1
- src/In.ProjectEKA.HipService/Discovery/Patient/Database/Migrations/DiscoveryContextModelSnapshot.cs  1
- src/In.ProjectEKA.HipService/Link/Patient/Database/Migrations/20200110085309_intial.Designer.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4833676&bid=16055942)

**Pull request merged**